### DISCUSSION TITLE
What is the difference between Autograd.backward() and backward in graph_executor.cc ?

### DISCUSSION THREAD

**Discussion by Johnny-dai-git** [apache/incubator-mxnet#20244] [General]
Hi, I am trying to understand the backward pass mechanism in MxNet. However, I came across a problem that I can not understand. While we are using autograd.backward(), the actually backward API in imperative.cc is not related to dependency Engine which means the backward pass do not need to push operator into the engine for dependency analysis. However, when we look into the graph_executor.cc , the backward API does push operator into the dependency engine. 
what is the difference between this two mechanism ? when should we use them ?

**Comment by TristonC** [comment]
All operations (forward as well as backward) are pushed into the engine. Why do you have to use backward in graph_execturo.cc?
